packages/labs/signals/src/lib/signal-watcher.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type {PropertyDeclaration, PropertyValueMap, ReactiveElement} from 'lit';
+import {Signal} from 'signal-polyfill';
+import {WatchDirective} from './watch.js';
+
+type ReactiveElementConstructor = abstract new (
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ...args: any[]
+) => ReactiveElement;
+
+export interface SignalWatcher extends ReactiveElement {
+  updateWatch(d: WatchDirective<unknown>): void;
+}
+
+/**
+ * Adds the ability for a LitElement or other ReactiveElement class to
+ * watch for access to signals during the update lifecycle and trigger a new
+ * update when signals values change.
+ */
+export function SignalWatcher<T extends ReactiveElementConstructor>(
+  Base: T
+): T {
+  abstract class SignalWatcher extends Base {
+    // Watcher.watch() doesn't dedupe so we need to track this ourselves.
+    private __isWatching = false;
+    private __watcherUpdate = false;
+    private __watcher = new Signal.subtle.Watcher(() => {
+      this.__watcherUpdate = true;
+      this.requestUpdate();
+      this.__watcherUpdate = false;
+    });
+    private __forceUpdateSignal = new Signal.State(0);
+    private __updateSignal = new Signal.Computed(() => {
+      this.__forceUpdateSignal.get();
+      super.performUpdate();
+    });
+    private __shouldRender = true;
+    private __pendingWatches = new Set<WatchDirective<unknown>>();
+
+    protected override performUpdate() {
+      // ReactiveElement.performUpdate() also does this check, so we want to
+      // bail early so we don't erroneously appear to not depend on any signals.
+      if (this.isUpdatePending === false) {
+        return;
+      }
+      if (this.__shouldRender) {
+        this.__updateSignal.get();
+      } else {
+        super.performUpdate();
+      }
+    }
+
+    protected override update(
+      changedProperties: PropertyValueMap<this> | Map<PropertyKey, unknown>
+    ): void {
+      if (this.__shouldRender) {
+        this.__shouldRender = false;
+        super.update(changedProperties);
+      } else {
+        this.__pendingWatches.forEach((d) => d.commmit());
+        this.__pendingWatches.clear();
+      }
+    }
+
+    override requestUpdate(
+      name?: PropertyKey | undefined,
+      oldValue?: unknown,
+      options?: PropertyDeclaration<unknown, unknown> | undefined
+    ): void {
+      this.__shouldRender = true;
+      if (!this.__watcherUpdate) {
+        this.__forceUpdateSignal?.set(this.__forceUpdateSignal.get() + 1);
+      }
+      super.requestUpdate(name, oldValue, options);
+    }
+
+    override connectedCallback(): void {
+      if (!this.__isWatching) {
+        this.__isWatching = true;
```

Comment:
Why is the flag needed? when would this happen when `__isWatching` is true? And I believe you can use `Signal.subtle.hasSources(this.__watcher)` if the state is needed (although an explicit flag in that case might be simpler).

Reply:
This logic has been simplified. I had to destroy and recreate watchers because of a polyfill bug.